language-support/java/bindings-rxjava/src/main/java/com/daml/ledger/rxjava/grpc/CommandsSubmission.java

```diff
@@ -0,0 +1,246 @@
+// Copyright (c) 2022 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.ledger.rxjava.grpc;
+
+import static java.util.Collections.emptyList;
+import static java.util.Collections.singletonList;
+import static java.util.Optional.empty;
+
+import com.daml.ledger.javaapi.data.codegen.HasCommands;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.List;
+import java.util.Optional;
+import org.checkerframework.checker.nullness.qual.NonNull;
+
+/**
+ * This class should be used to build a valid submission. It provides {@link #create(String, String, List)}
+ * for initial creation and extra builder methods for additional parameters
+ * e.g {@link #withActAs(List)}, {@link #withWorkflowId(String)} e.t.c
+ *
+ * Usage:
+ * <pre>
+ *   var params = CommandsSubmission.create("applicationId", "commandId", commands)
+ *                                   .withAccessToken("token")
+ *                                   .withParty("party")
+ *                                   .with...
+ * <pre/>
+ */
+public class CommandsSubmission {
+  private String applicationId;
+  private String commandId;
+  private List<@NonNull ? extends HasCommands> commands;
+
+  private String workflowId;
+  private List<@NonNull String> actAs;
+  private List<@NonNull String> readAs;
+  private Optional<Instant> minLedgerTimeAbs;
+  private Optional<Duration> minLedgerTimeRel;
+  private Optional<Duration> deduplicationTime;
+  private Optional<String> accessToken;
+
+  private CommandsSubmission(
+      String workflowId,
+      String applicationId,
+      String commandId,
+      List<@NonNull String> actAs,
+      List<@NonNull String> readAs,
+      Optional<Instant> minLedgerTimeAbs,
+      Optional<Duration> minLedgerTimeRel,
+      Optional<Duration> deduplicationTime,
+      List<@NonNull ? extends HasCommands> commands,
+      Optional<String> accessToken) {
+    this.workflowId = workflowId;
+    this.applicationId = applicationId;
+    this.commandId = commandId;
+    this.actAs = actAs;
+    this.readAs = readAs;
+    this.minLedgerTimeAbs = minLedgerTimeAbs;
+    this.minLedgerTimeRel = minLedgerTimeRel;
+    this.deduplicationTime = deduplicationTime;
+    this.commands = commands;
+    this.accessToken = accessToken;
+  }
+
+  public static CommandsSubmission create(
+      String applicationId, String commandId, List<@NonNull ? extends HasCommands> commands) {
+    return new CommandsSubmission(
+        "",
+        applicationId,
+        commandId,
+        emptyList(),
+        emptyList(),
+        empty(),
+        empty(),
+        empty(),
+        commands,
+        empty());
+  }
+
+  public String getWorkflowId() {
+    return workflowId;
+  }
+
+  public String getApplicationId() {
+    return applicationId;
+  }
+
+  public String getCommandId() {
+    return commandId;
+  }
+
+  public List<String> getActAs() {
```

Comment:
For this and other getters that return lists: if you don't want users to mutate what you return you need to return frozen collections or make defensive copies.